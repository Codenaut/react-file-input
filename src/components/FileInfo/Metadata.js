import React from 'react';
import PropTypes from 'prop-types';
import filesize from 'filesize';

import '../../styles/FileInfo.scss';

const Metadata = ({ extension, name, size, type }) => {
  const formattedSize = filesize(size, { separator: ',' });

  return (
    <div className="brainhub-file-info__metadata">
      <div className="brainhub-file-info__metadata__info">
        <span className="brainhub-file-info__metadata__info_name">{name}</span>{extension ? `.${extension}`: ''} ({formattedSize}) {type}
      </div>
    </div>
  );
};

Metadata.defaultProps = {
  extension: '',
  type: '',
};

Metadata.propTypes = {
  extension: PropTypes.string,
  name: PropTypes.string.isRequired,
  type: PropTypes.string,
  size: PropTypes.number.isRequired,
};

export default Metadata;

