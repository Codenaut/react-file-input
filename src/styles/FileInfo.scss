.brainhub-file-info {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-columns: 0.1fr 0.9fr;
  grid-template-areas: "image metadata";
  width: 100%;
  margin: 2px;
}
.brainhub-file-info__metadata {
  grid-area: metadata;
  display: flex;
  font-size: 0.8em;
  width: 100%;
  margin: 5px;
}

.brainhub-file-info__metadata__info{
  text-align: left;
  margin: auto;
  margin-left: 5px;
  margin-right: 0px;
}

.brainhub-file-info__metadata__info_name {
  font-weight: bold;
}

.brainhub-file-info__thumbnail {
  grid-area: image;
  width: 100%;
}

.brainhub-file-info__image {
  max-width: 100%;
  max-height: 100%;
  display: block;
  margin: auto;
}
