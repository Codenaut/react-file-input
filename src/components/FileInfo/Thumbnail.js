import React from 'react';
import Proptypes from 'prop-types';

import '../../styles/FileInfo.scss';

const Thumbnail = ({ children }) => (
  <div className="brainhub-file-info__thumbnail">
    {children}
  </div>
);

Thumbnail.propTypes = {
  children: Proptypes.node.isRequired,
};

export default Thumbnail;
