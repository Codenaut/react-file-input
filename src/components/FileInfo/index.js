import React from 'react';
import PropTypes from 'prop-types';

import Metadata from './Metadata';
import Thumbnail from './Thumbnail';

import '../../styles/FileInfo.scss';

const FileInfo = ({ file, image, metadataComponent, thumbnailComponent }) => {
  const MetadataClass = metadataComponent || Metadata;
  const ThumbnailClass = thumbnailComponent || Thumbnail;

  return (
    <div className="brainhub-file-info">
      <MetadataClass name={file.filename} size={file.size} extension={file.extension} type={file.mimeType}/>
      { (image && (
        <ThumbnailClass>
          <img className="brainhub-file-info__image" src={image}/>
        </ThumbnailClass>
      ))
      || null }
    </div>
  );
};

FileInfo.default = {
  image: '',
  metadataComponent: null,
  thumbnailComponent: null,
};

FileInfo.propTypes = {
  file: PropTypes.instanceOf(File).isRequired,
  image: PropTypes.string,
  metadataComponent: PropTypes.func,
  thumbnailComponent: PropTypes.func,
};

export default FileInfo;
