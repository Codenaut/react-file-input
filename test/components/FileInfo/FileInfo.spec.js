/* eslint-disable react/jsx-no-bind, react/prop-types */
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import FileInfo from '../../../src/components/FileInfo';
import Metadata from '../../../src/components/FileInfo/Metadata';
import Thumbnail from '../../../src/components/FileInfo/Thumbnail';

const metadataComponent = () => (<div className="customMetadata" />);
const thumbnailComponent = ({ children }) => (<div className="customThumbnail">{children}</div>);

const file = new File(['test.spec.js'], 'Test.spec.js', { type: 'test/mock' });
file.filename = 'Test';
file.extension = 'spec.js';
file.mimeType = file.type;

const defaultProps = {
  file,
};

const setup = (props = {}) => {
  const fileInfo = shallow(<FileInfo {...defaultProps} {...props} />);

  return {
    fileInfo,
  };
};

describe('components', () => {
  describe('FileInfo', () => {
    it('should render only a Metadata component', () => {
      const { fileInfo } = setup();

      expect(fileInfo.find(Metadata)).toHaveLength(1);
      expect(fileInfo.find(Thumbnail)).toHaveLength(0);
    });
    it('should render a Thumbnail if image is present', () => {
      const { fileInfo } = setup({ image: 'imageSource' });

      expect(fileInfo.find(Metadata)).toHaveLength(1);
      expect(fileInfo.find(Thumbnail)).toHaveLength(1);
    });
    it('should render a Thumbnail if image is present', () => {
      const { fileInfo } = setup({ image: 'imageSource', metadataComponent, thumbnailComponent });

      expect(fileInfo.find(metadataComponent)).toHaveLength(1);
      expect(fileInfo.find(thumbnailComponent)).toHaveLength(1);
      expect(fileInfo.find(Metadata)).toHaveLength(0);
      expect(fileInfo.find(Thumbnail)).toHaveLength(0);
    });
    it('should match exact snapshot', () => {
      const tree = renderer.create(
        <div>
          <FileInfo {...defaultProps}/>
        </div>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with custom class if defined', () => {
      const className = 'custom_class';
      const { fileInfo } = setup({ className });

      expect(fileInfo.hasClass(className)).toBeTruthy();
      expect(fileInfo.hasClass('brainhub-file-info')).toBeTruthy();
    });

    it('should match exact snapshot with custom components', () => {
      const tree = renderer.create(
        <div>
          <FileInfo
            {...defaultProps}
            image="testImage"
            metadataComponent={metadataComponent}
            thumbnailComponent={thumbnailComponent}
          />
        </div>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
