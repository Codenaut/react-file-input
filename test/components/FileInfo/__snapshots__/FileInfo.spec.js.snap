// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components FileInfo should match exact snapshot 1`] = `
<div>
  <div
    className="brainhub-file-info"
  >
    <div
      className="brainhub-file-info__metadata"
    >
      <div
        className="brainhub-file-info__metadata__info"
      >
        <span
          className="brainhub-file-info__metadata__info_name"
        >
          Test
        </span>
        .spec.js
         (
        12 B
        ) 
        test/mock
      </div>
    </div>
  </div>
</div>
`;

exports[`components FileInfo should match exact snapshot with custom components 1`] = `
<div>
  <div
    className="brainhub-file-info"
  >
    <div
      className="customMetadata"
    />
    <div
      className="customThumbnail"
    >
      <img
        className="brainhub-file-info__image"
        src="testImage"
      />
    </div>
  </div>
</div>
`;
